pkg/controller/controller.go

```diff
@@ -919,3 +948,69 @@ func validateIsolationGroups(cluster *myspec.M3DBCluster) error {
 	}
 	return nil
 }
+
+// shouldUpdateStatefulSet checks if a StatefulSet should be updated. If the Image or
+// ConfigMapName fields in the ClusterSpec don't match what's in the StatefulSet then
+// this function will update the provided StatefulSet to match the desired state and
+// return true.
+func shouldUpdateStatefulSet(spec myspec.ClusterSpec, sts *appsv1.StatefulSet) (bool, error) {
```

Comment:
Image and config updates are definitely the most common case we want to cover, so from that perspective this looks good. There are definitely other cases where the cluster CRD changes and we want to update the statefulsets, and I'm wondering how to address that here. Two thoughts:

1. Any idea what happens if we try to `DeepEqual` the generated spec and the current spec on the cluster? I'm guessing the Kubernetes API makes some sort of transformations that make that not totally possible, but I'm curious.

2. Assuming `DeepEqual` doesn't work, what do you think about a second annotation that will trigger an `Update` regardless of whether image and config has changed?

Reply:
I've always had difficulties using `DeepEqual` in the past because the API server sets default (non-zero) values for so many fields. Based on [this `kubebuilder` issue](https://github.com/kubernetes-sigs/kubebuilder/issues/592) though, it seems that we could use [`DeepDerivative`](https://pkg.go.dev/k8s.io/apimachinery/third_party/forked/golang/reflect#Equalities.DeepDerivative) instead since it "is similar to DeepEqual except that unset fields in a1 are ignored ". That issue also mentions [`github.com/banzaicloud/k8s-objectmatcher`](https://github.com/banzaicloud/k8s-objectmatcher) as another alternative if `DeepDerivative` doesn't work for us. If neither package is sufficient though I'm not opposed to having a `force-update` annotation or something similar which will always trigger an update.